### PR TITLE
Feature bulk import

### PR DESCRIPTION
**Pull request by bielu**
Hi Dave,
I was working on a feature for that package which will allow importing a list of redirects. 
Everything was done in cooperation with @ismailmayat. 
Here are a few screenshots of the functionality.
image
image (1)

I hope it will help not only for us.

### CODE REVIEW

File: Source/Our.Umbraco.RedirectsViewer/Controllers/RedirectsApiController.cs [hunk 1]
```diff
@@ -262,5 +249,76 @@ private string ValidateUrl(RedirectSave redirect)
...
+            Directory.CreateDirectory(uploadFolder);
+            var provider = new CustomMultipartFormDataStreamProvider(uploadFolder);
+
+            var result = await Request.Content.ReadAsMultipartAsync(provider);
+
+            var file = result.FileData[0];
+            var path = file.LocalFileName;
+            var ext = path.Substring(path.LastIndexOf('.')).ToLower();
+
+            if (ext != ".csv" && ext != ".xlsx")
```
[the remark this answers is not in the document]
Comment (bielu) [reply Source/Our.Umbraco.RedirectsViewer/Controllers/RedirectsApiController.cs]: @ismailmayat extracted.

File: Source/Our.Umbraco.RedirectsViewer/Controllers/RedirectsApiController.cs [hunk 2]
```diff
@@ -262,5 +249,76 @@ private string ValidateUrl(RedirectSave redirect)
...
+
+                    break;
+            }
+                
+            var response = Newtonsoft.Json.JsonConvert.SerializeObject(importer.Import(redirectsFile,result.FormData.Get("clientId")).StatusImportItems);
+          
+            return Request.CreateResponse(HttpStatusCode.OK, response);
+        }
+
+        public class CustomMultipartFormDataStreamProvider : MultipartFormDataStreamProvider
```
[the remark this answers is not in the document]
Comment (bielu) [reply Source/Our.Umbraco.RedirectsViewer/Controllers/RedirectsApiController.cs]: @ismailmayat extracted to separate class file.

File: Source/Our.Umbraco.RedirectsViewer/Import/Csv/CsvCellList.cs [hunk 3]
```diff
@@ -0,0 +1,93 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Skybrud.Umbraco.Redirects.Import.Csv {
```
[the remark this answers is not in the document]
Comment (bielu) [reply Source/Our.Umbraco.RedirectsViewer/Import/Csv/CsvCellList.cs]: Namespaces adjusted.

File: Source/Our.Umbraco.RedirectsViewer/Controllers/RedirectsApiController.cs [hunk 4]
```diff
@@ -262,5 +249,76 @@ private string ValidateUrl(RedirectSave redirect)
...
+            
+            var importer = new RedirectsImporterService(_redirectService,_domainService);
+            
+            IRedirectsFile redirectsFile;
+
+            switch (ext)
+            {   
+             
+                default:
+                    var csvFile = new CsvRedirectsFile(new RedirectPublishedContentFinder(UmbracoContext.ContentCache))
```
[the remark this answers is not in the document]
Comment (bielu) [reply Source/Our.Umbraco.RedirectsViewer/Controllers/RedirectsApiController.cs]: @ismailmayat  it is not that easy as the implementation of Umbraco content Cache is not possible to be injected from Umbraco DI.

File: Source/Our.Umbraco.RedirectsViewer/Web/App_Plugins/RedirectsViewer/controllers/ImportDialog.js [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        return Upload.upload({
+            file: $scope.file,
+            fields: {
+                'clientId': client
+            },
+            url:    apiUrl + "/import",
+        
+        }).then(function (response) {
+            if (response) 
+                //var fileName = response.data;
```
[the remark this answers is not in the document]
Comment (bielu) [reply Source/Our.Umbraco.RedirectsViewer/Web/App_Plugins/RedirectsViewer/controllers/ImportDialog.js]: @ismailmayat cleaned

File: Source/Our.Umbraco.RedirectsViewer/Web/App_Plugins/RedirectsViewer/controllers/ImportDialog.js [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    $scope.$on("filesSelected", function (event, args) {
+        if (args.files.length <= 0 || $scope.processing) {
+            $scope.file = null;
+            return;
+        }
+        $scope.noFile = false;
+        var file = args.files[0];
+        var extension = file.name.substring(file.name.lastIndexOf(".") + 1, file.name.length).toLowerCase();
+        console.log(extension);
+        if (extension !== "csv" && extension !== "xlsx") {
```
[the remark this answers is not in the document]
Comment (bielu) [reply Source/Our.Umbraco.RedirectsViewer/Web/App_Plugins/RedirectsViewer/controllers/ImportDialog.js]: @ismailmayat extracted.

File: Source/Our.Umbraco.RedirectsViewer/Web/App_Plugins/RedirectsViewer/controllers/ImportDialog.js [hunk 7]
```diff
@@ -0,0 +1,121 @@
+﻿angular.module('umbraco').controller('Our.Umbraco.RedirectsViewer.ImportDialog.Controller', function ($rootScope, $scope, $http, notificationsService, localizationService, editorService, Upload, $timeout) {
+
```
[the remark this answers is not in the document]
Comment (bielu) [reply Source/Our.Umbraco.RedirectsViewer/Web/App_Plugins/RedirectsViewer/controllers/ImportDialog.js]: @ismailmayat  using strict now :)

File: Source/Our.Umbraco.RedirectsViewer/Web/App_Plugins/RedirectsViewer/controllers/ImportDialog.js [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+        
+        }).then(function (response) {
+            if (response) 
+                //var fileName = response.data;
+
+                $scope.error = false;
+                $scope.processed = true;
+                $scope.processing = false;
+
+                //var file = new Blob([data], { type: 'text/csv' });
```
[the remark this answers is not in the document]
Comment (bielu) [reply Source/Our.Umbraco.RedirectsViewer/Web/App_Plugins/RedirectsViewer/controllers/ImportDialog.js]: @ismailmayat cleaned

File: Source/Our.Umbraco.RedirectsViewer/Controllers/RedirectsApiController.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (bielu) [reply Source/Our.Umbraco.RedirectsViewer/Controllers/RedirectsApiController.cs]: @ismailmayat solved

**Comment by dawoe** [comment]
Hi @bielu 

I will not be merging this in. This is not functionality that I envisioned for this package. It's called redirect viewer. I know it allows creating and deleting, but having a bulk import is not what I want to be in this package.

Also I guess you want this to import urls from a old site to create redirects for them. In my opinion this is better done on IIS level. Performance wise it's better to let the server handle it than the software.

Also if merge this in it will become my responsibility to maintain, and seeing my current workload this is not something I have time for.

But you can always release this as a seperate package. 

Dave

**Pull request closed without merging**